File: backend-python/app/services/mercadolivre.py
import asyncio
import re
import time
import httpx
import logging
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)

# Cache de token OAuth (module-level, compartilhado entre chamadas)
_token_cache: dict = {"token": None, "expires_at": 0.0}

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "pt-BR,pt;q=0.9",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
}

_MLB_RE = re.compile(r"(MLB\d+)")


def make_affiliate_url(url: str, tool_id: str) -> str:
    """Retorna a URL do produto com parâmetros de afiliado ML."""
    sep = "&" if "?" in url else "?"
    return f"{url}{sep}matt_tool={tool_id}&matt_source=affiliate"


# ---------------------------------------------------------------------------
# API oficial
# ---------------------------------------------------------------------------

async def _get_token(client_id: str, client_secret: str) -> str | None:
    """Obtém bearer token via client_credentials. Reutiliza cache enquanto válido."""
    now = time.time()
    if _token_cache["token"] and now < _token_cache["expires_at"] - 60:
        return _token_cache["token"]
    try:
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.post(
                "https://api.mercadolibre.com/oauth/token",
                data={
                    "grant_type": "client_credentials",
                    "client_id": client_id,
                    "client_secret": client_secret,
                },
            )
        if r.status_code != 200:
            logger.warning("ml.token_error", extra={"status": r.status_code})
            return None
        data = r.json()
        _token_cache["token"] = data["access_token"]
        _token_cache["expires_at"] = now + data.get("expires_in", 21600)
        logger.info("ml.token_renewed")
        return _token_cache["token"]
    except Exception as e:
        logger.error("ml.token_failed", extra={"error": str(e)})
        return None


async def _search_api(
    query: str, min_val: float, max_val: float, token: str
) -> list[dict]:
    """Busca via API oficial do ML."""
    try:
        async with httpx.AsyncClient(timeout=20) as c:
            r = await c.get(
                "https://api.mercadolibre.com/sites/MLB/search",
                params={
                    "q": query,
                    "price_min": int(min_val),
                    "price_max": int(max_val),
                    "limit": 20,
                    "sort": "price_asc",
                },
                headers={"Authorization": f"Bearer {token}"},
            )
        if r.status_code != 200:
            logger.warning("ml.api_error", extra={"status": r.status_code})
            return []
        results = []
        for item in r.json().get("results", []):
            price = float(item.get("price", 0))
            if not (min_val <= price <= max_val):
                continue
            results.append({
                "title": item.get("title", ""),
                "price": price,
                "url": item.get("permalink", ""),
                "image_url": item.get("thumbnail"),
                "source": "mercadolivre",
            })
        return results
    except Exception as e:
        logger.error("ml.api_failed", extra={"error": str(e)})
        return []


# ---------------------------------------------------------------------------
# Scraping HTML (fallback)
# ---------------------------------------------------------------------------

def _build_url(query: str, min_val: float, max_val: float) -> str:
    slug = re.sub(r"\s+", "-", query.strip().lower())
    slug = re.sub(r"[^\w-]", "", slug)
    return (
        f"https://lista.mercadolivre.com.br/{slug}"
        f"_PriceRange_{int(min_val)}-{int(max_val)}_NoIndex_True"
    )


def _parse(html: str, min_val: float, max_val: float) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    cards = soup.select("div.poly-card--grid-card")
    results = []

    for card in cards:
        title_el = card.select_one(".poly-component__title")
        if not title_el:
            continue

        title = title_el.get_text(strip=True)

        m = _MLB_RE.search(str(card))
        if not m:
            continue
        url = f"https://www.mercadolivre.com.br/p/{m.group(1)}"

        price_int_el = card.select_one(".andes-money-amount__fraction")
        if not price_int_el:
            continue

        try:
            int_part = price_int_el.get_text(strip=True).replace(".", "")
            price_cents_el = card.select_one(".andes-money-amount__cents")
            cents_part = price_cents_el.get_text(strip=True) if price_cents_el else "00"
            price = float(f"{int_part}.{cents_part}")
        except ValueError:
            continue

        if not (min_val <= price <= max_val):
            continue

        img_el = card.select_one(".poly-card__portada img")
        # data-src é a imagem real; src pode ser placeholder de lazy-load
        img_url = (img_el.get("data-src") or img_el.get("src")) if img_el else None

        if not any(r["url"] == url for r in results):
            results.append({
                "title": title,
                "price": price,
                "url": url,
                "image_url": img_url,
                "source": "mercadolivre",
            })

        if len(results) >= 20:
            break

    return results


# ---------------------------------------------------------------------------
# Entrypoint público
# ---------------------------------------------------------------------------

async def search(
    query: str,
    min_val: float,
    max_val: float,
    client_id: str | None = None,
    client_secret: str | None = None,
) -> list[dict]:
    # Tenta API oficial se credenciais configuradas
    if client_id and client_secret:
        token = await _get_token(client_id, client_secret)
        if token:
            results = await _search_api(query, min_val, max_val, token)
            if results:
                logger.info("ml.api_results", extra={"query": query, "count": len(results)})
                return results
            logger.warning("ml.api_empty_fallback", extra={"query": query})

    # Fallback: scraping HTML (com retry + backoff)
    url = _build_url(query, min_val, max_val)
    for attempt in range(3):
        try:
            async with httpx.AsyncClient(
                timeout=20, headers=HEADERS, follow_redirects=True
            ) as client:
                resp = await client.get(url)
            if resp.status_code == 200:
                results = _parse(resp.text, min_val, max_val)
                logger.info("ml.scraping_results", extra={"query": query, "count": len(results)})
                return results
            if resp.status_code == 429:
                wait = 5 * (attempt + 1)
                logger.warning("ml.rate_limited", extra={"attempt": attempt + 1, "wait_s": wait})
                await asyncio.sleep(wait)
                continue
            logger.warning("ml.scraping_http_error", extra={"status": resp.status_code})
            break
        except Exception as e:
            if attempt < 2:
                logger.warning("ml.scraping_retry", extra={"attempt": attempt + 1, "error": str(e)})
                await asyncio.sleep(3 * (attempt + 1))
            else:
                logger.error("ml.scraping_failed", extra={"error": str(e)})
    return []

File: backend-go/internal/scrapers/mercadolivre.go
package scrapers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"snatcher/backendv2/internal/pipeline"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/PuerkitoBio/goquery"
)

const mlBaseURL = "https://lista.mercadolivre.com.br"

type MLScraper struct {
	client       *http.Client
	clientID     string
	clientSecret string
	tokenMu      sync.Mutex
	token        string
	tokenExpiry  time.Time
}

func NewMLScraper(clientID, clientSecret string) *MLScraper {
	return &MLScraper{
		client:       &http.Client{Timeout: 20 * time.Second},
		clientID:     clientID,
		clientSecret: clientSecret,
	}
}

func (s *MLScraper) Search(ctx context.Context, query string, minVal, maxVal float64) ([]pipeline.Item, error) {
	if s.clientID != "" && s.clientSecret != "" {
		items, err := s.searchAPI(ctx, query, minVal, maxVal)
		if err == nil {
			return items, nil
		}
	}
	return s.searchHTML(ctx, query, minVal, maxVal)
}

func (s *MLScraper) Provider() string { return "mercadolivre" }

// searchAPI usa a API oficial do ML com OAuth2 client_credentials.
func (s *MLScraper) searchAPI(ctx context.Context, query string, minVal, maxVal float64) ([]pipeline.Item, error) {
	token, err := s.getToken(ctx)
	if err != nil {
		return nil, err
	}

	endpoint := fmt.Sprintf(
		"https://api.mercadolibre.com/sites/MLB/search?q=%s&price=%g-%g&limit=20",
		url.QueryEscape(query), minVal, maxVal,
	)
	req, _ := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("ml api status %d", resp.StatusCode)
	}

	var body struct {
		Results []struct {
			Title     string `json:"title"`
			Price     float64 `json:"price"`
			Permalink string `json:"permalink"`
			Thumbnail string `json:"thumbnail"`
		} `json:"results"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, err
	}

	out := make([]pipeline.Item, 0, len(body.Results))
	for _, r := range body.Results {
		if r.Price < minVal || r.Price > maxVal {
			continue
		}
		out = append(out, pipeline.Item{
			Title:    r.Title,
			Price:    r.Price,
			URL:      r.Permalink,
			ImageURL: r.Thumbnail,
			Source:   "mercadolivre",
		})
	}
	return out, nil
}

func (s *MLScraper) getToken(ctx context.Context) (string, error) {
	s.tokenMu.Lock()
	defer s.tokenMu.Unlock()
	if time.Now().Before(s.tokenExpiry) && s.token != "" {
		return s.token, nil
	}

	body := url.Values{
		"grant_type":    {"client_credentials"},
		"client_id":     {s.clientID},
		"client_secret": {s.clientSecret},
	}
	req, _ := http.NewRequestWithContext(ctx, "POST",
		"https://api.mercadolibre.com/oauth/token",
		strings.NewReader(body.Encode()))
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := s.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var tok struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&tok); err != nil {
		return "", err
	}
	s.token = tok.AccessToken
	s.tokenExpiry = time.Now().Add(time.Duration(tok.ExpiresIn-60) * time.Second)
	return s.token, nil
}

// searchHTML usa scraping da página de listagem como fallback.
func (s *MLScraper) searchHTML(ctx context.Context, query string, minVal, maxVal float64) ([]pipeline.Item, error) {
	searchURL := fmt.Sprintf("%s/%s", mlBaseURL, url.PathEscape(query))
	req, _ := http.NewRequestWithContext(ctx, "GET", searchURL, nil)
	req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36")
	req.Header.Set("Accept-Language", "pt-BR,pt;q=0.9")

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("ml html status %d", resp.StatusCode)
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}

	var items []pipeline.Item
	doc.Find(".ui-search-result__wrapper").Each(func(_ int, sel *goquery.Selection) {
		title := strings.TrimSpace(sel.Find(".ui-search-item__title").Text())
		priceText := strings.TrimSpace(sel.Find(".andes-money-amount__fraction").First().Text())
		priceText = strings.ReplaceAll(priceText, ".", "")
		priceText = strings.ReplaceAll(priceText, ",", ".")
		price, _ := strconv.ParseFloat(priceText, 64)
		link, _ := sel.Find("a.ui-search-link").First().Attr("href")
		img, _ := sel.Find("img.ui-search-result-image__element").First().Attr("data-src")
		if img == "" {
			img, _ = sel.Find("img.ui-search-result-image__element").First().Attr("src")
		}

		if title == "" || price == 0 || link == "" {
			return
		}
		if price < minVal || price > maxVal {
			return
		}
		// Limpa parâmetros de tracking do link
		if u, err := url.Parse(link); err == nil {
			u.RawQuery = ""
			link = u.String()
		}
		items = append(items, pipeline.Item{
			Title:    title,
			Price:    price,
			URL:      link,
			ImageURL: img,
			Source:   "mercadolivre",
		})
	})
	return items, nil
}

// readAll é helper para leitura de body (não usado mas mantido para referência)
func readAll(r io.Reader) ([]byte, error) {
	return io.ReadAll(io.LimitReader(r, 2<<20))
}
